lib/charms/dns_authority/v0/dns_authority.py

```diff
@@ -0,0 +1,206 @@
+# Copyright 2025 Canonical Ltd.
+# Licensed under the Apache2.0. See LICENSE file in charm source for details.
+
+"""Library to manage the integration with an authoritative DNS charm."""
+
+# The unique Charmhub library identifier, never change it
+LIBID = "b2697d5736cdc98807860010f911d65d"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = ["pydantic>=2"]
+
+# pylint: disable=wrong-import-position
+import json
+import logging
+import re
+import typing
+
+import ops
+import pydantic
+
+DEFAULT_RELATION_NAME = "dns-authority"
+
+logger = logging.getLogger(__name__)
+
+
+class DNSAuthorityRelationData(pydantic.BaseModel):
+    """Pydantic model representing the DNS authority relation data.
+
+    Attrs:
+        addresses: list of IP addresses of the DNS authority servers
+        zones: list of DNS zone names uplon which the servers have authority
+    """
+
+    addresses: list[pydantic.IPvAnyAddress]
+    zones: list[str]
+
+    @pydantic.model_validator(mode="before")
+    @classmethod
+    def ensure_uniqueness(cls, values: dict[str, typing.Any]) -> dict[str, typing.Any]:
+        """Ensure addresses and zones are unique.
+
+        Args:
+            values: input values for the model
+
+        Returns:
+            validated values
+        """
+        if "addresses" in values and isinstance(values["addresses"], list):
+            values["addresses"] = list(dict.fromkeys(values["addresses"]))
+        else:
+            raise ValueError("Incorrect input for 'addresses'")
+        if "zones" in values and isinstance(values["zones"], list):
+            values["zones"] = list(dict.fromkeys(values["zones"]))
+        return values
+
+    @pydantic.field_validator("zones")
+    @classmethod
+    def validate_zones(cls, zones: list[str]) -> list[str]:
+        """Validate the input zones.
+
+        Args:
+            zones: list of DNS zone names
+
+        Returns:
+            A list of DNS zone names
+
+        Raises:
+            ValueError: if the input has an invalid DNS zone name
+
+        Our main references are RFC 1034 and 2181
+        RFC 1034: https://datatracker.ietf.org/doc/html/rfc1034#section-3.1
+        RFC 2181: https://datatracker.ietf.org/doc/html/rfc2181#section-11
+        """
+        # rfc2181: "Any binary string whatever can be used as the label of any resource record",
+        # But we still want to reduce the available space.
+        # A common regex for a label is the following:
+        label_pattern = r"^[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9]$|^[a-zA-Z0-9]$"
```

Comment:
Underscores missing

Reply:
Underscores are actually not allowed in hostnames. The confusion comes from resource records attributes that can have them, lke TXT records